File: fries-leetcode-training/06-binary-tree/09-invert-binary-tree.cpp
/**
 * LeetCode 226. 翻转二叉树
 * https://leetcode.cn/problems/invert-binary-tree/
 * 
 * 给你一棵二叉树的根节点root，翻转这棵二叉树，并返回其根节点。
 * 
 * 递归
 * 
 * 时间复杂度：O(n)
 * 空间复杂度：O(h) h为树的高度
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if (!root) {
            return nullptr;
        }
        
        // 翻转左右子树
        TreeNode* temp = root->left;
        root->left = root->right;
        root->right = temp;
        
        // 递归翻转子树
        invertTree(root->left);
        invertTree(root->right);
        
        return root;
    }
};

// 测试函数
#include <iostream>
using namespace std;

void testInvertTree() {
    Solution solution;
    
    // 创建测试二叉树: [4,2,7,1,3,6,9]
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(9);
    
    TreeNode* inverted = solution.invertTree(root);
    
    cout << "翻转后的根节点值: " << inverted->val << endl;
    cout << "左子树根节点值: " << (inverted->left ? to_string(inverted->left->val) : "null") << endl;
    cout << "右子树根节点值: " << (inverted->right ? to_string(inverted->right->val) : "null") << endl;
}

int main() {
    testInvertTree();
    return 0;
}

File: fries-leetcode-training/06-binary-tree/09-invert-binary-tree.py
"""
LeetCode 226. 翻转二叉树
https://leetcode.cn/problems/invert-binary-tree/

给你一棵二叉树的根节点root，翻转这棵二叉树，并返回其根节点。

递归

时间复杂度：O(n)
空间复杂度：O(h) h为树的高度
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def invert_tree(root):
    """
    翻转二叉树 - 递归法
    
    Args:
        root: 二叉树根节点
        
    Returns:
        翻转后的二叉树根节点
    """
    if not root:
        return None
    
    # 翻转左右子树
    root.left, root.right = root.right, root.left
    
    # 递归翻转子树
    invert_tree(root.left)
    invert_tree(root.right)
    
    return root


def test_invert_tree():
    """测试函数"""
    # 创建测试二叉树: [4,2,7,1,3,6,9]
    #     4
    #    / \
    #   2   7
    #  / \ / \
    # 1  3 6  9
    
    root = TreeNode(4)
    root.left = TreeNode(2)
    root.right = TreeNode(7)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(9)
    
    inverted = invert_tree(root)
    
    def print_tree(node):
        if not node:
            return
        print(node.val, end=" ")
        print_tree(node.left)
        print_tree(node.right)
    
    print("翻转后的前序遍历:")
    print_tree(inverted)
    print()


if __name__ == "__main__":
    test_invert_tree()
